stf/src/saros/stf/client/StfTestCase.java

```diff
@@ -27,600 +26,754 @@
+     * {@link #createTestThread}.
+     *
+     * @param timeout
+     *            timeout in milliseconds to wait for the termination of all
+     *            test threads
+     * @throws IllegalStateException
+     *             if there are still test threads running
+     */
+    public static void terminateTestThreads(long timeout) {
+        for (EclipseTestThread thread : currentTestThreads)
+            thread.interrupt();
+
+        if (Util.joinAll(timeout,
+            currentTestThreads.toArray(new EclipseTestThread[0]))) {
+            currentTestThreads.clear();
+            return;
+        }
+
+        for (Iterator<EclipseTestThread> it = currentTestThreads.iterator(); it
+            .hasNext();) {
+
+            EclipseTestThread thread = it.next();
 
-      EclipseTestThread thread = it.next();
+            if (!thread.isAlive())
+                it.remove();
+        }
 
-      if (!thread.isAlive()) it.remove();
+        if (!currentTestThreads.isEmpty())
+            throw new IllegalStateException(
+                currentTestThreads.size() + " test threads are still alive");
     }
 
-    if (!currentTestThreads.isEmpty())
-      throw new IllegalStateException(currentTestThreads.size() + " test threads are still alive");
-  }
+    @SuppressWarnings("deprecation")
+    private static void checkAndStopRunningTestThreads(boolean abortAllTests) {
 
-  @SuppressWarnings("deprecation")
-  private static void checkAndStopRunningTestThreads(boolean abortAllTests) {
+        if (currentTestThreads.isEmpty() && !StfTestCase.abortAllTests)
+            return;
 
-    if (currentTestThreads.isEmpty() && !StfTestCase.abortAllTests) return;
+        assert lastTestClass != null;
 
-    assert lastTestClass != null;
+        IllegalStateException exception = new IllegalStateException(
+            "could not terminate all test threads for test class "
+                + lastTestClass.getName());
 
-    IllegalStateException exception =
-        new IllegalStateException(
-            "could not terminate all test threads for test class " + lastTestClass.getName());
+        if (!abortAllTests)
+            throw exception;
 
-    if (!abortAllTests) throw exception;
+        // do it the hard way and give up
 
-    // do it the hard way and give up
+        StfTestCase.abortAllTests = true;
 
-    StfTestCase.abortAllTests = true;
+        for (EclipseTestThread thread : currentTestThreads)
+            thread.stop();
 
-    for (EclipseTestThread thread : currentTestThreads) thread.stop();
+        currentTestThreads.clear();
 
-    currentTestThreads.clear();
+        throw exception;
+    }
 
-    throw exception;
-  }
+    /**
+     * Checks if the test for establishing a session between two users passed
+     */
+    public static boolean checkIfShare2UsersSequentiallySucceeded() {
+        for (String failedTest : allFailedTests) {
+            if (failedTest.contains("Share2UsersSequentiallyTest")) {
```

Comment:
I am deeply sorry but this is not a good solution because it is error prone by design.

Reply:
With the help of a TestWatcher, the name of the test class is written in allFailedTests if at least one of the tests in this test class failed.  You mean it is error-prone because you cannot rely on the TestWatcher? Could you please give more details?